src/tools/rbd_mirror/image_sync/ImageCopyRequest.cc

```diff
@@ -210,6 +217,48 @@ void ImageCopyRequest<I>::handle_object_copy(int r) {
 }
 
 template <typename I>
+void ImageCopyRequest<I>::send_update_sync_point() {
+  assert(m_lock.is_locked());
+
+  if (m_canceled || m_ret_val < 0) {
+    return;
+  }
+
+  update_progress("UPDATE_SYNC_POINT");
+
+  m_client_meta_copy = *m_client_meta;
+  m_updating_sync_point = true;
+  m_sync_point->object_number = m_object_no - 1;
+
+  CephContext *cct = m_local_image_ctx->cct;
+  ldout(cct, 20) << ": sync_point=" << *m_sync_point << dendl;
+
+  bufferlist client_data_bl;
+  librbd::journal::ClientData client_data(*m_client_meta);
+  ::encode(client_data, client_data_bl);
+
+  Context *ctx = create_context_callback<
+    ImageCopyRequest<I>, &ImageCopyRequest<I>::handle_update_sync_point>(
+      this);
+  m_journaler->update_client(client_data_bl, ctx);
+}
+
+template <typename I>
+void ImageCopyRequest<I>::handle_update_sync_point(int r) {
+  CephContext *cct = m_local_image_ctx->cct;
+  ldout(cct, 20) << ": r=" << r << dendl;
+
+  if (r < 0) {
+    *m_client_meta = m_client_meta_copy;
+    lderr(cct) << ": failed to update client data: " << cpp_strerror(r)
+               << dendl;
+  }
+
+  Mutex::Locker l(m_lock);
+  m_updating_sync_point = false;
```

Comment:
Need to handle the cases where the state machine completes / cancels while there is an in-flight sync point update. In such a case, you cannot complete the state machine until the sync point callback has been invoked / processed.